### PR TITLE
Integrate DA Forcing Module

### PR DESCRIPTION
**Pull request by shorvath-noaa**
Integration of new DA forcing module into t-route. All DA options (streamflow nudging, persistence reservoirs, and RFC reservoirs) can now work with the BMI modules.

## Additions

- Use of reservoir_rfc_da.py functions within mc_reach.pyx if running t-route via BMI functions
- Creation of reservoir_rfc_df and reservoir_rfc_param_df dataframes in the data_assimilation object
- Tracking of rfc DA indexing and update times
- Loading all DA data on initialization of DAforcing module, performing QA/QC, then pass dataframes of timeslice file info and timeseries file info directly to t-route model
- Read RFC reservoir gage-lake crosswalk info from a csv file. This is a temporary solution until RFC gages are included in the new hydrofabric
- 

## Removals

-

## Changes

- New method of converting gage info from hydrofabric 'network' table into gage-segment crosswalk and gage-lake crosswalk.
- When an individual lake can be either USGS or USACE, default to USGS
- Adjust logic in reservoir_rfc_da() to determine whether or not to advance to the next index

## Testing

1. Passed a 5 day simulation via BMI functions
2. Confirmed by plotting that an individual RFC reservoir's outflows are matching gage info (have not exhaustively checked all reservoirs)

## Screenshots
Reservoir 167299819 with gage ID GRRA4
image

## Notes

-

## Todos

- 

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

File: src/troute-network/troute/HYFeaturesNetwork.py [hunk 1]
```diff
@@ -431,64 +431,77 @@ def preprocess_data_assimilation(self, network):
...
+                sort_values(['usace_gage_id','hydroseq']).groupby('usace_lake_id').
+                last().
+                drop('hydroseq', axis=1)
+            )
+            
+            # Set waterbody types if DA is turned on:
+            streamflow_nudging = self.data_assimilation_parameters.get('streamflow_da',{}).get('streamflow_nudging',False)
+            usgs_da = self.data_assimilation_parameters.get('reservoir_da',{}).get('reservoir_persistence_usgs',False)
+            usace_da = self.data_assimilation_parameters.get('reservoir_da',{}).get('reservoir_persistence_usace',False)
+            rfc_da = self.waterbody_parameters.get('rfc',{}).get('reservoir_rfc_forecasts',False)
```
Comment (kumdonoaa) [inline src/troute-network/troute/HYFeaturesNetwork.py:484]: It seems like test_conus_AnA.yaml doesn't have reservoir_rfc_forecasts parameter. Instead, test_DA_config.yaml has it.  Don't we need to use self.data_assimilation_paramters instead then?

Comment (shorvath-noaa) [reply src/troute-network/troute/HYFeaturesNetwork.py:484]: Yeah, you're right. RFC parameters used to be under waterbody_parameters, but I moved it when working on the new troute.config module into data_assimilation_parameters where I think it makes more sense. I'll update this.

File: src/troute-routing/troute/routing/fast_reach/mc_reach.pyx [hunk 2]
```diff
@@ -653,4 +748,4 @@ cpdef object compute_network_structured(
...
\ No newline at end of file
+    return np.asarray(data_idx, dtype=np.intp)[fill_index_mask], output.reshape(output.shape[0], -1)[fill_index_mask], 0, (np.asarray([data_idx[usgs_position_i] for usgs_position_i in usgs_positions]), np.asarray(lastobs_times), np.asarray(lastobs_values)), (usgs_idx, usgs_update_time-((timestep-1)*dt), usgs_prev_persisted_ouflow, usgs_prev_persistence_index, usgs_persistence_update_time-((timestep-1)*dt)), (usace_idx, usace_update_time-((timestep-1)*dt), usace_prev_persisted_ouflow, usace_prev_persistence_index, usace_persistence_update_time-((timestep-1)*dt)), output_upstream.reshape(output.shape[0], -1)[fill_index_mask], (rfc_idx, rfc_update_time-((timestep-1)*dt), rfc_timeseries_idx)
```
Comment (kumdonoaa) [inline src/troute-routing/troute/routing/fast_reach/mc_reach.pyx:751]: Are the new returned variables used in computation when returned in compute.py?  I guess computation over the given entire time loop takes place in mc_reach.pyx, so these variables values never gets used for any next time steps?

Comment (shorvath-noaa) [reply src/troute-routing/troute/routing/fast_reach/mc_reach.pyx:751]: Yes, they are passed to the data_assimilation object to update the rfc_param_df for the next loop.

File: src/troute-routing/troute/routing/fast_reach/reservoir_RFC_da.py [hunk 3]
```diff
@@ -229,8 +229,8 @@ def reservoir_RFC_da(use_RFC, time_series, timeseries_idx, total_counts, routing
     Notes
     -----
     ''' 
-    if use_RFC and (current_time+routing_period)<=rfc_forecast_persist_seconds:
-        if (current_time+routing_period) >= update_time and timeseries_idx<total_counts:
+    if use_RFC and (current_time)<=rfc_forecast_persist_seconds:
```
[the remark this answers is not in the document]
Comment (shorvath-noaa) [reply src/troute-routing/troute/routing/fast_reach/reservoir_RFC_da.py:232]: Adding routing_period to this made the reservoir module update the index 1 timestep too early.

File: src/troute-routing/troute/routing/compute.py [hunk 4]
```diff
@@ -254,6 +298,7 @@ def compute_nhd_routing_v02(
     waterbody_type_specified,
     subnetwork_list,
     flowveldepth_interorder = {},
+    from_files = True,
```
Comment (kumdonoaa) [inline src/troute-routing/troute/routing/compute.py:301]: At line 306, there is an error saying "ValueError: could not convert string to float: '01010000,01010000,01010000,01010000' .  When I checked param_df's gages column, I found

(Pdb) test.gages
key
302    01010000,01010000,01010000,01010000
Name: gages, dtype: object

I think a single stream take a single gage at most, but it seems weird.

**Review by kumdonoaa — APPROVED** [review]

**Pull request merged**